### DISCUSSION TITLE
Access to Handler instance

### DISCUSSION THREAD

**Discussion by DavideCordella** [LuckyPennySoftware/MediatR#817] [Q&A]
Hi I'm new to MediatR package, I'm implementing a custom behavior for the pipeline and I need to get access to the running handler instance in order to run some methods contained in it according to some value in the request.
At the moment I can see this is not possible. Am I correct? Any chance to get this feature in the future?

Many thanks.

**Comment by jbogard** [comment]
No, and this is intentionally by design. If you need some common behavior between multiple dependencies, extract that common logic into its own service and have each class take that common dependency.

**Reply by DavideCordella** [reply]
Hi thanks for your answer, the behavior (method signature) is common to all handlers but it must have different implementations for each one. Can you advise? Thanks

**Reply by jbogard** [reply]
Different for each one?? I don't know your use case but it sounds like some sort of strategy pattern. I can't answer more without seeing your specific use case.

**Reply by DavideCordella** [reply]
you hit the point, but that's only required in the pipeline behavior

**Reply by jbogard** [reply]
Then you have two options:
1. Refactor your handlers in a way that does not force a behavior to call a method on a handler instance
2. Refactor your handlers away from MediatR into something that more appropriately fits your scenario

I'd explore number 2 first, pretend that MediatR doesn't exist. What interfaces would you define? Are behaviors now being tucked away that's not explicit and obvious? The new design might make that behavior an explicit part of the caller.

Without seeing actual code I can't really help more.
